packages/ember-toucan-core/src/components/form/controls/combobox.ts

```diff
@@ -0,0 +1,537 @@
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import { assert } from '@ember/debug';
+import { action } from '@ember/object';
+import { guidFor } from '@ember/object/internals';
+import { next } from '@ember/runloop';
+import { isEqual as emberIsEqual } from '@ember/utils';
+
+import { offset, size } from '@floating-ui/dom';
+
+import OptionComponent, {
+  selector as optionComponentSelector,
+} from '../../../-private/components/form/controls/combobox/option';
+import Chevron from '../../../-private/icons/chevron';
+
+import type { Middleware as VelcroMiddleware } from '@floating-ui/dom';
+import type { WithBoundArgs } from '@glint/template';
+
+export type Option = string | Record<string, unknown> | undefined;
+
+export interface ToucanFormComboboxControlComponentSignature<
+  OPTION extends Option
+> {
+  Args: {
+    /**
+     * A CSS class to add to this component's content container.
+     * Commonly used to specify a `z-index`.
+     */
+    contentClass?: string;
+
+    /**
+     * Sets the input to an errored-state via styling.
+     */
+    hasError?: boolean;
+
+    /**
+     * Sets the `disabled` attribute of the input.
+     */
+    isDisabled?: boolean;
+
+    /**
+     * Sets the `readonly` attribute of the input.
+     */
+    isReadOnly?: boolean;
+
+    /**
+     * A string to display when there are no results after the user's filter.
+     */
+    noResultsText?: string;
+
+    /**
+     * Called when the user makes a selection.
+     * It is called with the selected option (derived from `@options`) as its only argument.
+     */
+    onChange?: (option: unknown) => void;
+
+    /**
+     * The function called when a user types into the combobox textbox, typically used to write custom filtering logic.
+     */
+    onFilter?: (input: string) => OPTION[];
+
+    /**
+     * `@options` forms the content of this component.
+     *
+     * To support a variety of data shapes, `@options` is typed as `unknown[]` and treated as though it were opaque.
+     * `@options` is simply iterated over then passed back to you as a block parameter (`select.option`).
+     */
+    options?: OPTION[];
+
+    /**
+     * When `@options` is an array of objects, `@selected` is also an object.
+     * The `@optionKey` is used to determine which key of the object should
+     * be used for both filtering and displayed the selected value in the
+     * textbox.
+     */
+    optionKey?: string;
```

Comment:
We maybe able to type this in a more strict (correct) way:

```suggestion
    optionKey?: OPTION extends Record<string, unknown> ? keyof OPTION : undefined;
```

I haven't tried this, so maybe it doesn't work or causes issues. But given...
* `optionsOfObjects = [{label: 'foo'}, {label: 'bar'}]`
* `optionsOfStrings = ['foo' 'bar']`

ideally TS would give us this feedback...
* `@options={{optionsOfObjects)} @optionKey="label"` ✅
* `@options={{optionsOfStrings}}` ✅
* `@options={{optionsOfObjects)} @optionKey="title"` ❌
* `@options={{optionsOfStrings)} @optionKey="label"` ❌
* `@options={{optionsOfObjects)}` ❌

Reply:
It works! This is great! Thanks a bunch, Simon!

<img width="769" alt="Screenshot 2023-07-14 at 8 36 29 AM" src="https://github.com/CrowdStrike/ember-toucan-core/assets/8069555/a0f58cde-9b64-41de-a586-adaba3c42f2a">